{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CTCF Recall and Precision"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_regions(bedfile):\n",
    "    return len(bedfile)\n",
    "\n",
    "def recall_precision(ctcf_bed, peak_bedgraph):\n",
    "    \"\"\"\n",
    "    Calculates recall and precision for CTCF binding sites.\n",
    "\n",
    "    Args:\n",
    "        ctcf_bed (str): Path to CTCF motif BED file.\n",
    "        peak_bedgraph (str): Path to peak bedgraph file.\n",
    "\n",
    "    Returns:\n",
    "        tuple: (recall, precision, total_peak_regions, peaks_with_ctcf)\n",
    "    \"\"\"\n",
    "    print(f\"peak_bedgraph: {peak_bedgraph}\")\n",
    "    ctcf_regions = BedTool(ctcf_bed)\n",
    "    peak_regions = BedTool(peak_bedgraph)\n",
    "    print(f\"Num peak regions: {num_regions(peak_regions)}\")\n",
    "\n",
    "    num_ctcf = num_regions(ctcf_regions)\n",
    "    print(f\"Number of CTCF regions: {num_ctcf}\")\n",
    "\n",
    "    # Overlap CTCF regions with peaks\n",
    "    overlapping_from_ctcf_regions = ctcf_regions.intersect(peak_regions, u=True)\n",
    "\n",
    "    num_overlapping_ctcf = num_regions(overlapping_from_ctcf_regions)\n",
    "    print(f\"Number of CTCF overlapping regions: {num_overlapping_ctcf}\")\n",
    "\n",
    "    recall = num_overlapping_ctcf / num_ctcf if num_ctcf > 0 else 0\n",
    "\n",
    "    # Overlap peaks with CTCF regions\n",
    "    peaks_with_ctcf = peak_regions.intersect(ctcf_regions, u=True)\n",
    "    num_peaks_with_ctcf = num_regions(peaks_with_ctcf)\n",
    "    print(f\"Number of peaks with CTCF: {num_peaks_with_ctcf}\")\n",
    "\n",
    "    precision = num_peaks_with_ctcf / num_regions(peak_regions) if num_regions(peak_regions) > 0 else 0\n",
    "\n",
    "    return recall, precision, num_regions(peak_regions), num_peaks_with_ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_recall_precision(\n",
    "    ctcf_bed,\n",
    "    before_bed,\n",
    "    after_bed,\n",
    "    outfile='ctcf_recall_precision_plot.png',\n",
    "    cell_line='GM12878',\n",
    "    sample='ENCFF355CYX',\n",
    "    fdr='0.1',\n",
    "    bar_labels=['Before', 'After'],\n",
    "    experiment_type='ChIP-Seq'\n",
    "):\n",
    "    \"\"\"\n",
    "    Plots recall and precision for CTCF binding sites before and after chrom-sig.\n",
    "\n",
    "    Args:\n",
    "        ctcf_bed (str): Path to CTCF motif BED file.\n",
    "        before_bed (str): Path to peak bedgraph file before chrom-sig.\n",
    "        after_bed (str): Path to peak bedgraph file after chrom-sig.\n",
    "        outfile (str): Output filename for the plot.\n",
    "        cell_line (str): Cell line name.\n",
    "        sample (str): Sample name.\n",
    "        fdr (str): FDR value.\n",
    "    \"\"\"\n",
    "\n",
    "    before_recall, before_precision, before_total_regions, before_ctcf_regions = recall_precision(ctcf_bed, before_bed)\n",
    "    after_recall, after_precision, after_total_regions, after_ctcf_regions = recall_precision(ctcf_bed, after_bed)\n",
    "\n",
    "    height = 0.35  # Bar height\n",
    "    labels = ['Recall', 'Precision']\n",
    "    x = np.arange(len(labels))  # [0, 1]\n",
    "\n",
    "    before_scores = [before_recall, before_precision]\n",
    "    after_scores = [after_recall, after_precision]\n",
    "\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    bars1 = plt.barh(x + height/2, before_scores, height, label=bar_labels[0], color='skyblue')\n",
    "    bars2 = plt.barh(x - height/2, after_scores, height, label=bar_labels[1], color='orange')\n",
    "    plt.xlim(0, 1)\n",
    "    plt.yticks(x, labels)\n",
    "    plt.xlabel('Score')\n",
    "    plt.title(f'Recall and Precision of CTCF Motifs {cell_line} {experiment_type} {sample} FDR {fdr}')\n",
    "    plt.legend()\n",
    "    plt.grid(axis='y')\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Add score labels and region counts to each bar\n",
    "    for i, bar in enumerate(bars1):\n",
    "        plt.text(bar.get_width() + 0.02, bar.get_y() + bar.get_height()/2, f\"{bar.get_width():.3f}\",\n",
    "                 va='center', ha='left', fontsize=10)\n",
    "        if i == 1:\n",
    "            plt.text(0.01, bar.get_y() + bar.get_height()/2 - 0.08,\n",
    "                     f\"Total regions: {before_total_regions}\", va='center', ha='left', fontsize=10)\n",
    "            plt.text(0.01, bar.get_y() + bar.get_height()/2 + 0.08,\n",
    "                     f\"CTCF regions: {before_ctcf_regions}\", va='center', ha='left', fontsize=10)\n",
    "    for i, bar in enumerate(bars2):\n",
    "        plt.text(bar.get_width() + 0.02, bar.get_y() + bar.get_height()/2, f\"{bar.get_width():.3f}\",\n",
    "                 va='center', ha='left', fontsize=10)\n",
    "        if i == 1:\n",
    "            plt.text(0.01, bar.get_y() + bar.get_height()/2 - 0.08,\n",
    "                     f\"Total regions: {after_total_regions}\", va='center', ha='left', fontsize=10)\n",
    "            plt.text(0.01, bar.get_y() + bar.get_height()/2 + 0.08,\n",
    "                     f\"CTCF regions: {after_ctcf_regions}\", va='center', ha='left', fontsize=10)\n",
    "\n",
    "    plt.savefig(outfile)\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctcf_bed = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CTCF_motifs_STORM_hg38_Ext4kbBoth.sorted.id.bed'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GM12878 4DNFI9U71IB4 CTCF CUTRUN at different FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_raw_SICER.bedgraph\n",
      "Num peak regions: 62176\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21323\n",
      "Number of peaks with CTCF: 23799\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bedgraph\n",
      "Num peak regions: 19233\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 14437\n",
      "Number of peaks with CTCF: 14198\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_raw_SICER.bedgraph\n",
      "Num peak regions: 62176\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21323\n",
      "Number of peaks with CTCF: 23799\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_chrom-sig_FDR0.2_pseudoRead_5000_SICER.bedgraph\n",
      "Num peak regions: 31957\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 18363\n",
      "Number of peaks with CTCF: 19041\n"
     ]
    }
   ],
   "source": [
    "# GM12878 CUT&RUN example\n",
    "before_bed = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_raw_SICER.bedgraph'\n",
    "after_bed_fdr1 = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bedgraph'\n",
    "after_bed_fdr2 = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_chrom-sig_FDR0.2_pseudoRead_5000_SICER.bedgraph'\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, before_bed, after_bed_fdr1,\n",
    "        outfile='plots/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_precision_recall_plot_FDR0.1.png',\n",
    "        cell_line='GM12878', sample='4DNFI9U71IB4', fdr='0.1'\n",
    "    )\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, before_bed, after_bed_fdr2,\n",
    "        outfile='plots/GM12878_CUTRUN_CTCF_4DNFI9U71IB4_precision_recall_plot_FDR0.2.png',\n",
    "        cell_line='GM12878', sample='4DNFI9U71IB4', fdr='0.2'\n",
    "    )\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recall and precision at varying SICER e-vals and top N peaks\n",
    "\n",
    "We compare recall and precision for peaks called using SICER at different e-vals compared to Chrom-Sig at FDR 0.1 and 5000 pseudoreads.  \n",
    "We compare recall and precision for the top N peaks called using SICER compared to Chrom-Sig at FDR 0.1 and 5000 pseudoreads.  \n",
    "\n",
    "#### GM12878 ENCFF355CYX CTCF ChIP-seq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e1500.scoreisland\n",
      "Num peak regions: 37271\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22090\n",
      "Number of peaks with CTCF: 22674\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e2000.scoreisland\n",
      "Num peak regions: 37804\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22145\n",
      "Number of peaks with CTCF: 22780\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e500.scoreisland\n",
      "Num peak regions: 34761\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21881\n",
      "Number of peaks with CTCF: 22262\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e750.scoreisland\n",
      "Num peak regions: 35592\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21951\n",
      "Number of peaks with CTCF: 22403\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_10000_peaks.bed\n",
      "Num peak regions: 10000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 9885\n",
      "Number of peaks with CTCF: 8910\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_1000_peaks.bed\n",
      "Num peak regions: 1000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 1291\n",
      "Number of peaks with CTCF: 903\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_15000_peaks.bed\n",
      "Num peak regions: 15000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 13906\n",
      "Number of peaks with CTCF: 13006\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_20000_peaks.bed\n",
      "Num peak regions: 20000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 17271\n",
      "Number of peaks with CTCF: 16608\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_5000_peaks.bed\n",
      "Num peak regions: 5000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 5385\n",
      "Number of peaks with CTCF: 4542\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# CTCF ChIP-Seq example\n",
    "\n",
    "SICER_DIR = '/nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/'\n",
    "e_vals = ['GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e1500.scoreisland',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e2000.scoreisland',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e500.scoreisland',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600_e750.scoreisland']\n",
    "\n",
    "top_peaks = ['GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_10000_peaks.bed',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_1000_peaks.bed',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_15000_peaks.bed',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_20000_peaks.bed',\n",
    "    'GM12878_CTCF_ChIP-seq_ENCFF355CYX_bam2bed.std-W200-G600.scoreisland_top_5000_peaks.bed']\n",
    "\n",
    "chrom_sig_bed = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_ChIP-seq_CTCF_ENCFF355CYX_chrom-sig_FDR0.1_pseudoRead5000_SICER.bed'\n",
    "# make outdir if it doesnt exist\n",
    "outdir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/plots/revisions/CTCF_CHIP-Seq_ENCFF355CYX'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "for file in e_vals:\n",
    "    e_val_bed = SICER_DIR + file\n",
    "    e_val = file.split('_e')[-1].split('.')[0]\n",
    "    plot_recall_precision(\n",
    "            ctcf_bed, e_val_bed, chrom_sig_bed,\n",
    "            outfile=f'{outdir}/e_val{e_val}_vs_Chrom-Sig_precision_recall_plot.png',\n",
    "            cell_line='GM12878', sample='ENCFF355CYX', fdr=0.1,\n",
    "            bar_labels=[f'e-val {e_val}', 'Chrom-Sig']\n",
    "    )\n",
    "    \n",
    "for file in top_peaks:\n",
    "    top_peak_bed = SICER_DIR + file\n",
    "    num_peaks = file.split('_top_')[-1].split('_peaks')[0]\n",
    "    plot_recall_precision(\n",
    "            ctcf_bed, top_peak_bed, chrom_sig_bed,\n",
    "            outfile=f'{outdir}/top_{num_peaks}_peaks_vs_Chrom-Sig_precision_recall_plot.png',\n",
    "            cell_line='GM12878', sample='ENCFF355CYX', fdr=0.1,\n",
    "            bar_labels=[f'Top {num_peaks} peaks', 'Chrom-Sig']\n",
    "    )\n",
    "\n",
    "\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GM12878 4DNFI2G71DR4 CTCF CUTRUN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e1500.scoreisland\n",
      "Num peak regions: 56726\n",
      "Num peak regions: 56726\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20950\n",
      "Number of CTCF overlapping regions: 20950\n",
      "Number of peaks with CTCF: 22772\n",
      "Number of peaks with CTCF: 22772\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e2000.scoreisland\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e2000.scoreisland\n",
      "Num peak regions: 58506\n",
      "Num peak regions: 58506\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21060\n",
      "Number of CTCF overlapping regions: 21060\n",
      "Number of peaks with CTCF: 23017\n",
      "Number of peaks with CTCF: 23017\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e500.scoreisland\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e500.scoreisland\n",
      "Num peak regions: 52258\n",
      "Num peak regions: 52258\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20652\n",
      "Number of CTCF overlapping regions: 20652\n",
      "Number of peaks with CTCF: 22111\n",
      "Number of peaks with CTCF: 22111\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e750.scoreisland\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e750.scoreisland\n",
      "Num peak regions: 53946\n",
      "Num peak regions: 53946\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20786\n",
      "Number of CTCF overlapping regions: 20786\n",
      "Number of peaks with CTCF: 22376\n",
      "Number of peaks with CTCF: 22376\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_10000_peaks.bed\n",
      "Num peak regions: 10000\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_10000_peaks.bed\n",
      "Num peak regions: 10000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 8554\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 8554\n",
      "Number of peaks with CTCF: 7512\n",
      "Number of peaks with CTCF: 7512\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_1000_peaks.bed\n",
      "Num peak regions: 1000\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_1000_peaks.bed\n",
      "Num peak regions: 1000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 1200\n",
      "Number of peaks with CTCF: 785\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 1200\n",
      "Number of peaks with CTCF: 785\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_15000_peaks.bed\n",
      "Num peak regions: 15000\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_15000_peaks.bed\n",
      "Num peak regions: 15000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 11639\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 11639\n",
      "Number of peaks with CTCF: 10669\n",
      "Number of peaks with CTCF: 10669\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_20000_peaks.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_20000_peaks.bed\n",
      "Num peak regions: 20000\n",
      "Num peak regions: 20000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 14036\n",
      "Number of CTCF overlapping regions: 14036\n",
      "Number of peaks with CTCF: 13288\n",
      "Number of peaks with CTCF: 13288\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_5000_peaks.bed\n",
      "Num peak regions: 5000\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_5000_peaks.bed\n",
      "Num peak regions: 5000\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 4833\n",
      "Number of peaks with CTCF: 3910\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 4833\n",
      "Number of peaks with CTCF: 3910\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n"
     ]
    }
   ],
   "source": [
    " # run motif on GM12878 CUT&RUN without chrom-sig at different SICER e-vals and compare to chrom-sig results\n",
    "CR_SICER = '/nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_CUTandRUN'\n",
    "\n",
    "CR_e_vals = ['GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e1500.scoreisland',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e2000.scoreisland',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e500.scoreisland',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600_e750.scoreisland']\n",
    "\n",
    "CR_top_peaks = ['GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_10000_peaks.bed',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_1000_peaks.bed',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_15000_peaks.bed',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_20000_peaks.bed',\n",
    "    'GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland_top_5000_peaks.bed']\n",
    "\n",
    "cr_outdir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/plots/revisions/CTCF_CUTRUN_4DNFI2G71DR4'\n",
    "os.makedirs(cr_outdir, exist_ok=True)\n",
    "chrom_sig_cr_bed = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/GM12878_CUTRUN_CTCF_4DNFI2G71DR4_chrom-sig_FDR0.1_pseudoRead_5000_SICER.bed'\n",
    "for file in CR_e_vals:\n",
    "    e_val_bed = CR_SICER + '/' + file\n",
    "    e_val = file.split('_e')[-1].split('.')[0]\n",
    "    plot_recall_precision(\n",
    "            ctcf_bed, e_val_bed, chrom_sig_cr_bed,\n",
    "            outfile=f'{cr_outdir}/GM12878_CR_4DNFI2G71DR4_e_val{e_val}_vs_Chrom-Sig_precision_recall_plot.png',\n",
    "            cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.1,\n",
    "            bar_labels=[f'e-val {e_val}', 'Chrom-Sig'],\n",
    "            experiment_type='CUT&RUN'\n",
    "        )\n",
    "\n",
    "for file in CR_top_peaks:\n",
    "    top_peak_bed = CR_SICER + '/' + file\n",
    "    num_peaks = file.split('_top_')[-1].split('_peaks')[0]\n",
    "    plot_recall_precision(\n",
    "            ctcf_bed, top_peak_bed, chrom_sig_cr_bed,\n",
    "            cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.1,\n",
    "            outfile=f'{cr_outdir}/GM12878_CR_4DNFI2G71DR4_top_{num_peaks}_peaks_vs_Chrom-Sig_precision_recall_plot.png',\n",
    "            bar_labels=[f'Top {num_peaks} peaks', 'Chrom-Sig'],\n",
    "            experiment_type='CUT&RUN'\n",
    "    )\n",
    "\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MACS2 vs SICER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = '/nfs/turbo/umms-minjilab/njgupta/chromsig'\n",
    "macs_dir = root_dir + '/MACS2_Results'\n",
    "macs_out_dir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/plots/revisions/CTCF_CHIP-Seq_ENCFF668ROZ_MACS2_comparison'\n",
    "\n",
    "os.makedirs(macs_out_dir, exist_ok=True)\n",
    "\n",
    "# CTCF ChIP-seq ENCFF668ROZ\n",
    "# original SICER and MACS2 results\n",
    "ori_sicer_roz = root_dir + '/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_bam2bed.std-W200-G600.scoreisland'\n",
    "ori_macs_wo_roz = macs_dir + '/ChIP-seq/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_summits.bed'\n",
    "ori_macs_w_ctrl_roz = macs_dir + '/With_control/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_IgG_summits.bed'\n",
    "\n",
    "# chrom-sig results\n",
    "CS_roz = root_dir + '/ChIP-seq_SE/GM12878_CTCF_ChIP-seq_ENCFF668ROZ/ENCFF668ROZ_EnrichTest_FDR_0.2_pseudoRead_5000/ENCFF668ROZ_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland'\n",
    "CS_macs_wo_roz = macs_dir + '/Post_chromsig/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_FDR_0.2_pseudoRead_5000_summits.bed'\n",
    "CS_macs_ctrl_roz = macs_dir + '/Post_chromsig_control/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_total_FDR_0.2_pseudoRead_5000_pass_pileup_IgG_summits.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_bam2bed.std-W200-G600.scoreisland\n",
      "Num peak regions: 43531\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22555\n",
      "Number of peaks with CTCF: 23823\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/ChIP-seq/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_summits.bed\n",
      "Num peak regions: 49171\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22590\n",
      "Number of peaks with CTCF: 24471\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_BASIC/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_bam2bed.std-W200-G600.scoreisland\n",
      "Num peak regions: 43531\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22555\n",
      "Number of peaks with CTCF: 23823\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/With_control/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_IgG_summits.bed\n",
      "Num peak regions: 49767\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22577\n",
      "Number of peaks with CTCF: 26425\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/ChIP-seq_SE/GM12878_CTCF_ChIP-seq_ENCFF668ROZ/ENCFF668ROZ_EnrichTest_FDR_0.2_pseudoRead_5000/ENCFF668ROZ_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland\n",
      "Num peak regions: 38435\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22466\n",
      "Number of peaks with CTCF: 23695\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/Post_chromsig/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_FDR_0.2_pseudoRead_5000_summits.bed\n",
      "Num peak regions: 58926\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22593\n",
      "Number of peaks with CTCF: 24492\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/ChIP-seq_SE/GM12878_CTCF_ChIP-seq_ENCFF668ROZ/ENCFF668ROZ_EnrichTest_FDR_0.2_pseudoRead_5000/ENCFF668ROZ_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland\n",
      "Num peak regions: 38435\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22466\n",
      "Number of peaks with CTCF: 23695\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/Post_chromsig_control/GM12878_CTCF_ChIP-seq_ENCFF668ROZ_total_FDR_0.2_pseudoRead_5000_pass_pileup_IgG_summits.bed\n",
      "Num peak regions: 51018\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22572\n",
      "Number of peaks with CTCF: 26644\n"
     ]
    }
   ],
   "source": [
    "plot_recall_precision(\n",
    "        ctcf_bed, ori_sicer_roz, ori_macs_wo_roz,\n",
    "        outfile=f'{macs_out_dir}/ENCFF668ROZ_orig_SICER_vs_MAC2_no_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='ENCFF668ROZ', fdr=0.2,\n",
    "        bar_labels=['Original SICER', 'Original MACS2 (no control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, ori_sicer_roz, ori_macs_w_ctrl_roz,\n",
    "        outfile=f'{macs_out_dir}/ENCFF668ROZ_orig_SICER_vs_MAC2_with_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='ENCFF668ROZ', fdr=0.2,\n",
    "        bar_labels=['Original SICER', 'Original MACS2 (with control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, CS_roz, CS_macs_wo_roz,\n",
    "        outfile=f'{macs_out_dir}/ENCFF668ROZ_chrom-sig_SICER_vs_MAC2_no_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='ENCFF668ROZ', fdr=0.2,\n",
    "        bar_labels=['Chrom-Sig SICER', 'Chrom-Sig MACS2 (no control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, CS_roz, CS_macs_ctrl_roz,\n",
    "        outfile=f'{macs_out_dir}/ENCFF668ROZ_chrom-sig_SICER_vs_MAC2_with_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='ENCFF668ROZ', fdr=0.2,\n",
    "        bar_labels=['Chrom-Sig SICER', 'Chrom-Sig MACS2 (with control)']\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CTCF CUT&RUN 4DNFI2G71DR4\n",
    "# original SICER and MACS2 results\n",
    "ori_sicer_dr4 = root_dir + '/SICER_Clusters/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland'\n",
    "ori_macs_wo_dr4 = macs_dir + '/CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_summits.bed'\n",
    "ori_macs_w_ctrl_dr4 = macs_dir + '/With_control/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_IgG_summits.bed'\n",
    "\n",
    "# chrom-sig results\n",
    "CS_dr4 = root_dir + '/SICER_Clusters_Pass_0_2/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland'\n",
    "CS_macs_wo_dr4 = macs_dir + '/Post_chromsig/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_FDR_0.2_pseudoRead_5000_summits.bed'\n",
    "CS_macs_ctrl_dr4 = macs_dir + '/Post_chromsig_control/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_total_FDR_0.2_pseudoRead_5000_pass_pileup_IgG_summits.bed'\n",
    "\n",
    "macs_out_dir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/plots/revisions/CTCF_CUTRUN_DR4_MACS2_comparison'\n",
    "\n",
    "os.makedirs(macs_out_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_Clusters/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland\n",
      "Num peak regions: 55251\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20867\n",
      "Number of peaks with CTCF: 22554\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/CUTandRUN/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_summits.bed\n",
      "Num peak regions: 56839\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20516\n",
      "Number of peaks with CTCF: 23559\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_Clusters/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_bam2bed.std-W200-G600.scoreisland\n",
      "Num peak regions: 55251\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20867\n",
      "Number of peaks with CTCF: 22554\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/With_control/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_IgG_summits.bed\n",
      "Num peak regions: 59239\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 20097\n",
      "Number of peaks with CTCF: 28272\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_Clusters_Pass_0_2/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland\n",
      "Num peak regions: 29346\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 17317\n",
      "Number of peaks with CTCF: 17896\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/Post_chromsig/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_FDR_0.2_pseudoRead_5000_summits.bed\n",
      "Num peak regions: 53957\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 18087\n",
      "Number of peaks with CTCF: 35172\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/SICER_Clusters_Pass_0_2/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_total_FDR_0.2_pseudoRead_5000_pass_pileup-W200-G600.scoreisland\n",
      "Num peak regions: 29346\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 17317\n",
      "Number of peaks with CTCF: 17896\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/njgupta/chromsig/MACS2_Results/Post_chromsig_control/GM12878_CUTandRUN_CTCF_4DNFI2G71DR4_total_FDR_0.2_pseudoRead_5000_pass_pileup_IgG_summits.bed\n",
      "Num peak regions: 63090\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 17951\n",
      "Number of peaks with CTCF: 43379\n"
     ]
    }
   ],
   "source": [
    "plot_recall_precision(\n",
    "        ctcf_bed, ori_sicer_dr4, ori_macs_wo_dr4,\n",
    "        outfile=f'{macs_out_dir}/4DNFI2G71DR4_orig_SICER_vs_MAC2_no_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.2, experiment_type='CUT&RUN',\n",
    "        bar_labels=['Original SICER', 'Original MACS2 (no control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, ori_sicer_dr4, ori_macs_w_ctrl_dr4,\n",
    "        outfile=f'{macs_out_dir}/4DNFI2G71DR4_orig_SICER_vs_MAC2_with_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.2, experiment_type='CUT&RUN',\n",
    "        bar_labels=['Original SICER', 'Original MACS2 (with control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, CS_dr4, CS_macs_wo_dr4,\n",
    "        outfile=f'{macs_out_dir}/4DNFI2G71DR4_chrom-sig_SICER_vs_MAC2_no_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.2, experiment_type='CUT&RUN',\n",
    "        bar_labels=['Chrom-Sig SICER', 'Chrom-Sig MACS2 (no control)']\n",
    "    )\n",
    "\n",
    "plot_recall_precision(\n",
    "        ctcf_bed, CS_dr4, CS_macs_ctrl_dr4,\n",
    "        outfile=f'{macs_out_dir}/4DNFI2G71DR4_chrom-sig_SICER_vs_MAC2_with_control_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample='4DNFI2G71DR4', fdr=0.2, experiment_type='CUT&RUN',\n",
    "        bar_labels=['Chrom-Sig SICER', 'Chrom-Sig MACS2 (with control)']\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AtacWorks Comparison\n",
    "\n",
    "We are comparing AtacWorks to Chrom-Sig for 2 CTCF ChIP-Seq replicate pairs and CTCF CUTRUN replicates.  \n",
    "AtacWorks was trained using the [pretrained bulk lowqual 20M reads model](https://api.ngc.nvidia.com/v2/models/org/nvidia/atac_bulk_lowqual_20m_20m/0.3/files?redirect=true&path=models/model.pth.tar) with the default parameters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Precision and Recall Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP1_ENCFF355CYX_sorted.bed\n",
      "Num peak regions: 445\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 140\n",
      "Number of peaks with CTCF: 153\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP1_ENCFF355CYX_sorted.bed\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 140\n",
      "Number of peaks with CTCF: 153\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP1_ENCFF355CYX_sorted.bed\n",
      "Num peak regions: 24872\n",
      "Num peak regions: 24872\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of CTCF overlapping regions: 19074\n",
      "Number of peaks with CTCF: 19132\n",
      "Number of peaks with CTCF: 19132\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed\n",
      "Num peak regions: 1635\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed\n",
      "Num peak regions: 1635\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 690\n",
      "Number of peaks with CTCF: 784\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 690\n",
      "Number of peaks with CTCF: 784\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed\n",
      "Num peak regions: 30250\n",
      "Num peak regions: 30250\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21401\n",
      "Number of CTCF overlapping regions: 21401\n",
      "Number of peaks with CTCF: 21902\n",
      "Number of peaks with CTCF: 21902\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed\n",
      "Num peak regions: 9350\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed\n",
      "Num peak regions: 9350\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 6683\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 6683\n",
      "Number of peaks with CTCF: 7708\n",
      "Number of peaks with CTCF: 7708\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed\n",
      "Num peak regions: 44918\n",
      "Num peak regions: 44918\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22560\n",
      "Number of CTCF overlapping regions: 22560\n",
      "Number of peaks with CTCF: 24351\n",
      "Number of peaks with CTCF: 24351\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed\n",
      "Num peak regions: 3991\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed\n",
      "Num peak regions: 3991\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 2579\n",
      "Number of peaks with CTCF: 3056\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 2579\n",
      "Number of peaks with CTCF: 3056\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed\n",
      "Num peak regions: 33005\n",
      "Num peak regions: 33005\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22057\n",
      "Number of CTCF overlapping regions: 22057\n",
      "Number of peaks with CTCF: 22849\n",
      "Number of peaks with CTCF: 22849\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_CUTRUN1_DR4_sorted.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_CUTRUN1_DR4_sorted.bed\n",
      "Num peak regions: 177485\n",
      "Number of CTCF regions: 22648\n",
      "Num peak regions: 177485\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 21797\n",
      "Number of CTCF overlapping regions: 21797\n",
      "Number of peaks with CTCF: 24288\n",
      "Number of peaks with CTCF: 24288\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_CUTRUN1_DR4_sorted.bed\n",
      "Num peak regions: 16774\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_CUTRUN1_DR4_sorted.bed\n",
      "Num peak regions: 16774\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 12831\n",
      "Number of peaks with CTCF: 12547\n",
      "Number of peaks with CTCF: 12547\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_CUTRUN2_IB4_sorted.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted/AW_GM12878_CTCF_CUTRUN2_IB4_sorted.bed\n",
      "Num peak regions: 251985\n",
      "Num peak regions: 251985\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 22291\n",
      "Number of CTCF overlapping regions: 22291\n",
      "Number of peaks with CTCF: 25153\n",
      "Number of peaks with CTCF: 25153\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_CUTRUN1_IB4_sorted.bed\n",
      "peak_bedgraph: /nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted/GM12878_CTCF_CUTRUN1_IB4_sorted.bed\n",
      "Num peak regions: 19233\n",
      "Num peak regions: 19233\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF regions: 22648\n",
      "Number of CTCF overlapping regions: 14437\n",
      "Number of CTCF overlapping regions: 14437\n",
      "Number of peaks with CTCF: 14198\n",
      "Number of peaks with CTCF: 14198\n"
     ]
    }
   ],
   "source": [
    "# atac works recall comparison\n",
    "AW_dir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/AtacWorks_sorted'\n",
    "CS_dir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/data/CS_sorted'\n",
    "aw_outdir = '/nfs/turbo/umms-minjilab/zapell/chrom-sig/plots/revisions/CTCF_AtacWorks_vs_Chrom-Sig'\n",
    "os.makedirs(aw_outdir, exist_ok=True)\n",
    "atac_works_beds = [\n",
    "        AW_dir + '/AW_GM12878_CTCF_ChIP1_ENCFF355CYX_sorted.bed',\n",
    "        AW_dir + '/AW_GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed',\n",
    "        AW_dir + '/AW_GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed',\n",
    "        AW_dir + '/AW_GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed',\n",
    "        AW_dir + '/AW_GM12878_CTCF_CUTRUN1_DR4_sorted.bed',\n",
    "        AW_dir + '/AW_GM12878_CTCF_CUTRUN2_IB4_sorted.bed'\n",
    "]\n",
    "\n",
    "cs_beds = [\n",
    "        CS_dir + '/GM12878_CTCF_ChIP1_ENCFF355CYX_sorted.bed',\n",
    "        CS_dir + '/GM12878_CTCF_ChIP1_ENCFF453EJM_sorted.bed',\n",
    "        CS_dir + '/GM12878_CTCF_ChIP2_ENCFF500UFQ_sorted.bed',\n",
    "        CS_dir + '/GM12878_CTCF_ChIP2_ENCFF668ROZ_sorted.bed',\n",
    "        CS_dir + '/GM12878_CTCF_CUTRUN1_DR4_sorted.bed',\n",
    "        CS_dir + '/GM12878_CTCF_CUTRUN1_IB4_sorted.bed'\n",
    "]\n",
    "\n",
    "samples = [\n",
    "        'ENCFF355CYX_ChIP1',\n",
    "        'ENCFF453EJM_ChIP1',\n",
    "        'ENCFF500UFQ_ChIP2',\n",
    "        'ENCFF668ROZ_ChIP2',\n",
    "        '4DNFI2G71DR4_CUTRUN1',\n",
    "        '4DNFI9U71IB4_CUTRUN1'\n",
    "]\n",
    "\n",
    "for i in range(len(atac_works_beds)):\n",
    "    aw_bed = atac_works_beds[i]\n",
    "    cs_bed = cs_beds[i]\n",
    "    sample = samples[i]\n",
    "    plot_recall_precision(\n",
    "        ctcf_bed, aw_bed, cs_bed,\n",
    "        outfile=f'{aw_outdir}/GM12878_CTCF_{sample}_AtacWorks_vs_Chrom-Sig_precision_recall_plot.png',\n",
    "        cell_line='GM12878', sample=sample, fdr='0.1',\n",
    "        bar_labels=['AtacWorks', 'Chrom-Sig'],\n",
    "        experiment_type='AtacWorks vs Chrom-Sig'\n",
    "        )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
